### DISCUSSION TITLE
物化视图支持UDAF吗

### DISCUSSION THREAD

**Discussion by eleven-f** [apache/doris#27023] [A - General / Q&A]
物化视图支持UDAF吗？

**Comment by ixzc** [comment]
yes

**Reply by eleven-f** [reply]
谢谢！我目前遇到个问题，我实现了一个字符串（纯数字）聚合求和的UDAF，在使用select的时候，能够正常出结果，但当我把这个查询用于建立物化视图是，总是失败，并且报错，能帮忙分析下吗，不胜感激！

建表语句：

`CREATE TABLE IF NOT EXISTS test.bsc_tx_records(
    from_addr CHAR(42),
    to_addr CHAR(42),
    value STRING COMMENT '交易额',
    fee STRING COMMENT '交易手续费',
    block_time datetime COMMENT '交易时间戳'
)
ENGINE=OLAP
PARTITION BY RANGE(block_time) ()
DISTRIBUTED BY HASH(block_time) BUCKETS AUTO            
PROPERTIES (
    "enable_duplicate_without_keys_by_default" = "true",   
    "dynamic_partition.enable" = "true",                  
    "dynamic_partition.create_history_partition" = "true", 
    "dynamic_partition.history_partition_num" = "36",     
    "dynamic_partition.time_unit" = "MONTH",             
    "dynamic_partition.prefix" = "bsc_tx",                
    "dynamic_partition.end" = "1"                          
);`

样例数据：

`INSERT INTO test.bsc_tx_records
VALUES ('addr1', 'addr2', '1000', '50', '2023-09-12 23:28:53'),
       ('addr1', 'addr2', '3000', '20', '2023-09-12 23:28:54')
;`

select结果：

![image](https://github.com/apache/doris/assets/30048177/4b9a900c-43f1-46a0-a7fe-2892976878b8)


物化视图：

`CREATE MATERIALIZED VIEW mv_bsc_tx_stats
AS SELECT from_addr, to_addr, MIN(block_time), MAX(block_time), udaf_sum(value), udaf_sum(fee)
FROM bsc_tx_records
GROUP BY from_addr, to_addr;`

创建物化视图错误信息：

`errCode = 2, detailMessage = rollup tasks failed on same tablet reach threshold [[ALTER], signature: 67431, backendId: 12037, tablet id: 67431, [ALTER], signature: 67431, backendId: 12038, tablet id: 67431], reason=task type: ALTER, status_code: INVALID_ARGUMENT, status_message: [(192.168.1.16)[INVALID_ARGUMENT][E33] DataTypeAggState function get failed, type=AggState(function_name=udaf_sum,result_is_nullable=true,arguments=[Nullable(String)]) 0. /root/src/doris-2.0/be/src/common/stack_trace.cpp:302: StackTrace::tryCapture() @ 0x000000000ba1f197 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 1. /root/src/doris-2.0/be/src/common/stack_trace.h:0: doris::get_stack_trace[abi:cxx11]() @ 0x000000000ba1d72d in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 2. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:173: doris::Exception::Exception(int, std::basic_string_view<char, std::char_traits<char> >) @ 0x000000000b4b996e in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 3. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:187: doris::Exception::Exception<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(int, std::basic_string_view<char, std::char_traits<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&) @ 0x000000000dc2f664 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 4. /root/src/doris-2.0/be/src/vec/data_types/data_type_agg_state.h:0: doris::vectorized::DataTypeAggState::DataTypeAggState(std::vector<std::shared_ptr<doris::vectorized::IDataType const>, std::allocator<std::shared_ptr<doris::vectorized::IDataType const> > >, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) @ 0x000000000dc40bff in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 5. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:187: decltype (::new ((void*)(0)) doris::vectorized::DataTypeAggState(std::declval<std::vector<std::shared_ptr<doris::vectorized::IDataType const>, std::allocator<std::shared_ptr<doris::vectorized::IDataType const> > >&>(), std::declval<bool>(), std::declval<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >())) std::construct_at<doris::vectorized::DataTypeAggState, std::vector<std::shared_ptr<doris::vectorized::IDataType const>, std::allocator<std::shared_ptr<doris::vectorized::IDataType const> > >&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(doris::vectorized::DataTypeAggState*, std::vector<std::shared_ptr<doris::vectorized::IDataType const>, std::allocator<std::shared_ptr<doris::vectorized::IDataType const> > >&, bool&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&) @ 0x000000000dc40718 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 6. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/move.h:206: doris::vectorized::DataTypeFactory::create_data_type(doris::TabletColumn const&, bool) @ 0x000000000dc39d24 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 7. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/shared_ptr_base.h:1291: doris::TabletSchema::create_block(bool) const @ 0x000000000b772150 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 8. /root/src/doris-2.0/be/src/olap/schema_change.cpp:0: doris::VSchemaChangeWithSorting::_inner_process(std::shared_ptr<doris::RowsetReader>, doris::RowsetWriter*, std::shared_ptr<doris::Tablet>, std::shared_ptr<doris::TabletSchema>) @ 0x000000000b6db125 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 9. /var/local/ldb-toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/shared_ptr_base.h:701: doris::SchemaChange::process(std::shared_ptr<doris::RowsetReader>, doris::RowsetWriter*, std::shared_ptr<doris::Tablet>, std::shared_ptr<doris::Tablet>, std::shared_ptr<doris::TabletSchema>) @ 0x000000000b6e9997 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 10. /root/src/doris-2.0/be/src/common/status.h:348: doris::SchemaChangeHandler::_convert_historical_rowsets(doris::SchemaChangeHandler::SchemaChangeParams const&) @ 0x000000000b6e197e in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 11. /root/src/doris-2.0/be/src/olap/schema_change.cpp:0: doris::SchemaChangeHandler::_do_process_alter_tablet_v2(doris::TAlterTabletReqV2 const&) @ 0x000000000b6def40 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 12. /root/src/doris-2.0/be/src/olap/schema_change.cpp:664: doris::SchemaChangeHandler::process_alter_tablet_v2(doris::TAlterTabletReqV2 const&) @ 0x000000000b6dd0ab in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 13. /root/src/doris-2.0/be/src/common/status.h:348: doris::EngineAlterTabletTask::execute() @ 0x000000000b77a518 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 14. /root/src/doris-2.0/be/src/common/status.h:432: doris::StorageEngine::execute_task(doris::EngineTask*) @ 0x000000000b64c739 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 15. /root/src/doris-2.0/be/src/common/status.h:348: doris::AlterTableTaskPool::_alter_tablet(doris::TAgentTaskRequest const&, long, doris::TTaskType::type, doris::TFinishTaskRequest*) @ 0x000000000b6952a1 in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 16. /root/src/doris-2.0/be/src/agent/task_worker_pool.cpp:1730: doris::AlterTableTaskPool::_alter_tablet_worker_thread_callback() @ 0x000000000b694f1a in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 17. /root/src/doris-2.0/be/src/util/threadpool.cpp:0: doris::ThreadPool::dispatch_thread() @ 0x000000000ba5bdaf in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 18. /var/local/ldb-toolchain/bin/../usr/include/pthread.h:562: doris::Thread::supervise_thread(void*) @ 0x000000000ba51d3c in /opt/apache-doris-2.0.2-bin-x64/be/lib/doris_be 19. ? @ 0x00007fc63dedeb43 in ? 20. ? @ 0x00007fc63df70a00 in ? ], backendId: Backend [id=12038, host=192.168.1.16, heartbeatPort=9050, alive=true, lastStartTime=2023-11-15 16:14:16, tags: {location=default}], signature: 67431`
